### ISSUE TITLE
(accessibility) Missing screen reader labels for standard buttons

### ISSUE THREAD

**Issue by Wildebrew** [angular/material.angular.io#693]
Standard buttons on the component demo pages are visually labeled with icons but have no screen reader labels.
If you run [Axe](http://www.deque.com) on a page with these buttons, such as the [checkbox demo](https://material.angular.io/components/checkbox/overview) it will report these issues as "buttons must have discernible text".

* The “view source” buttons have no screen reader label
* The “Edit in Stack Blitz” buttons have no screen reader label.
* The “Select a theme” buttons have no screen rader label.

To solve, add `aria-label="the screen reader label"` on the button elements.
e.g.

```
<button _ngcontent-teq-c501="" mat-icon-button="" type="button" aria-label="View source" class="mat-icon-button mat-button-base" 
cdk-describedby-host="" style="touch-action: none; user-select: none; -webkit-user-drag: none; -webkit-tap-highlight-color: rgba(0, 0, 0, 0);"><span class="mat-button-wrapper"><mat-icon _ngcontent-teq-c501="" aria-hidden="true" role="img" class="mat-icon notranslate material-icons mat-icon-no-color"><svg _ngcontent-teq-c501="" xmlns="http://www.w3.org/2000/svg" width="100%" height="100%" viewBox="0 0 24 24" fit="" preserveAspectRatio="xMidYMid meet" focusable="false"><path _ngcontent-teq-c501="" fill="none" d="M0 0h24v24H0V0z"></path><path _ngcontent-teq-c501="" d="M9.4 16.6L4.8 12l4.6-4.6L8 6l-6 6 6 6 1.4-1.4zm5.2 0l4.6-4.6-4.6-4.6L16 6l6 6-6 6-1.4-1.4z"></path></svg></mat-icon></span>
	<div matripple="" class="mat-button-ripple mat-ripple mat-button-ripple-round"></div>
	<div class="mat-button-focus-overlay"></div>
</button>
```

Alternatively, replace the existing `aria-describedby` attribute with `aria-labelledby` (assuming there is a valid target element, an element whose `id` attribute matches the value of the `aria-labelledby` attribute and whose text content is a suitable screen reader label for the button.

e.g.
`aria-labelledby="cdk-describedby-message-1"`

### The difference between aria-labelledby and aria-describedby

Every operable control has to have an accessible name, this is the text that, for instance, a screen reader would announce when the element is focused or that a speech recognition user would use to select the control.

You can assign an accessible name using one of 4 ways:
* Using `aria-labelledby` with a valid id reference
* Using `aria-label` with a string
* Using native labeling mechanism (for a `<button>` that's the text inside the element, for a text input it would be the text in an associated `<label>` element
* Using the `title` attribute (don't do that)

If there are multiple sources of an element's accessible name, the browser/screen reader will prioritize based on the above list. If there is an `aria-labelledby` attribute on the control and if the element it references has the text "Foo", the accessible name of the control will be foo.
This is true even if you add `aria-label="bar"` on the same element or if it has the visible text "Google".

When a control needs additional information (such as a tooltip, operating instructions or an error message) that information is put in its accessible description attribute (by the browser/operating system. You can specify the accessible description in two ways:
* Using `aria-describedby` with a valid id reference
* Using the `title` attribute

An accessible description is not consistently announced by screen readers and its announcement can be altered through user configuration settings.

In other words,
* An operable control has to have an accessible name, but it does not have to have an accessible description
* An accessible description is not a substitute for an accessible name

See the [accessible name and description section of the ARIA Authoring Practices Guide](https://www.w3.org/TR/wai-aria-practices-1.1/#names_and_descriptions) for more information.

**Comment by Splaktar** [comment]
Thank you for the issue report. The general issue was already reported in https://github.com/angular/material.angular.io/issues/671, but I'll keep this one open since it includes a lot of additional information.

Both the View Source and Edit in StackBlitz buttons use a `matTooltip` which applies the `aria-describedby`, but you are correct in that this is not sufficient.

**Comment by Splaktar** [comment]
PR https://github.com/angular/material.angular.io/pull/697 fixes this issue.
